File: agar-clone/server-go/aoi.go
package main

import "math"

// aoi.go — uniform spatial hash (AoiCell=1024) for Area-of-Interest queries and
// the cell/food/virus/eject broad-phase used by the eat passes. Rebuilt each
// tick from the live entity maps.

type gridEntry struct {
	x, y float64
	ref  interface{}
}

type Grid struct {
	cell int
	m    map[int64][]gridEntry
}

func newGrid() *Grid {
	return &Grid{cell: AoiCell, m: make(map[int64][]gridEntry)}
}

func cellKey(cx, cy int) int64 {
	return int64(cx)*100000 + int64(cy)
}

func (g *Grid) clear() {
	// reuse allocation where possible
	g.m = make(map[int64][]gridEntry, len(g.m))
}

func (g *Grid) cellOf(x, y float64) (int, int) {
	return int(math.Floor(x / float64(g.cell))), int(math.Floor(y / float64(g.cell)))
}

func (g *Grid) insert(x, y float64, ref interface{}) {
	cx, cy := g.cellOf(x, y)
	k := cellKey(cx, cy)
	g.m[k] = append(g.m[k], gridEntry{x: x, y: y, ref: ref})
}

// queryCircle returns every entry whose cell overlaps the square bounding
// (x,y,radius). Coarse: caller does the precise circle/overlap test.
func (g *Grid) queryCircle(x, y, radius float64, out []gridEntry) []gridEntry {
	out = out[:0]
	minx, miny := g.cellOf(x-radius, y-radius)
	maxx, maxy := g.cellOf(x+radius, y+radius)
	for cx := minx; cx <= maxx; cx++ {
		for cy := miny; cy <= maxy; cy++ {
			if bucket, ok := g.m[cellKey(cx, cy)]; ok {
				out = append(out, bucket...)
			}
		}
	}
	return out
}

File: agar-clone/server/aoi.js
"use strict";

// =============================================================================
// aoi.js — Spatial hash grid for Area-of-Interest queries and broad-phase collision.
//          Cuadricula hash espacial para consultas de Area de Interes y colision de fase amplia.
//
// Key concepts / Conceptos clave:
//   - Uniform grid with configurable cell size (default 1024 world units) / Cuadricula uniforme con tamano de celda configurable (por defecto 1024 unidades de mundo)
//   - insert(x,y,ref) places entities into hash buckets / insert(x,y,ref) coloca entidades en cubetas hash
//   - queryCircle(x,y,r) returns coarse candidates overlapping a bounding square / queryCircle(x,y,r) retorna candidatos gruesos que solapan un cuadrado delimitador
//   - Rebuilt every tick from live entity maps; reused by both eat-pass and viewport AoI / Se reconstruye cada tick desde mapas de entidades vivos; reutilizada por el pase de comida y AoI del viewport
// =============================================================================

// ---------------------------------------------------------------------------
// aoi.js — uniform spatial hash (AOI_CELL=1024) for Area-of-Interest queries
// and the cell/food/virus/eject broad-phase used by the eat passes.
//
// Rebuilt each tick from the live entity maps. Cheap for a few thousand bodies.
// (Reused verbatim in pattern from bullseye.io/server/aoi.js.)
// ---------------------------------------------------------------------------

const C = require("./constants.js");

function cellKey(cx, cy) {
  return cx * 100000 + cy;
}

class Grid {
  constructor(cell = C.AOI_CELL) {
    this.cell = cell;
    this.map = new Map(); // key -> array of { x, y, ref }
  }

  clear() {
    this.map.clear();
  }

  _cellOf(x, y) {
    return [Math.floor(x / this.cell), Math.floor(y / this.cell)];
  }

  insert(x, y, ref) {
    const [cx, cy] = this._cellOf(x, y);
    const key = cellKey(cx, cy);
    let bucket = this.map.get(key);
    if (!bucket) {
      bucket = [];
      this.map.set(key, bucket);
    }
    bucket.push({ x, y, ref });
  }

  // Return every entry whose cell overlaps the square bounding (x,y,radius).
  // Coarse: caller does the precise circle/overlap test.
  queryCircle(x, y, radius) {
    const out = [];
    const min = this._cellOf(x - radius, y - radius);
    const max = this._cellOf(x + radius, y + radius);
    for (let cx = min[0]; cx <= max[0]; cx++) {
      for (let cy = min[1]; cy <= max[1]; cy++) {
        const bucket = this.map.get(cellKey(cx, cy));
        if (bucket) for (const e of bucket) out.push(e);
      }
    }
    return out;
  }
}

module.exports = { Grid };
